txnkv/transaction/txn.go

```diff
@@ -638,6 +665,34 @@ func (txn *KVTxn) InitPipelinedMemDB() error {
 	return nil
 }
 
+func (txn *KVTxn) throttle() {
+	if txn.writeThrottleRatio >= 1 || txn.writeThrottleRatio < 0 {
+		logutil.BgLogger().Error(
+			"[pipelined dml] invalid write speed ratio",
+			zap.Float64("writeThrottleRatio", txn.writeThrottleRatio),
+			zap.Uint64("session", txn.committer.sessionID),
+			zap.Uint64("startTS", txn.startTS),
+		)
+		return
+	}
+
+	expectedFlushMs := txn.flushBatchDurationEWMA.Value()
+	// T_sleep / (T_sleep + T_flush) = writeThrottleRatio
+	sleepMs := int(txn.writeThrottleRatio / (1.0 - txn.writeThrottleRatio) * expectedFlushMs)
+	metrics.TiKVPipelinedFlushThrottleSecondsHistogram.Observe(float64(sleepMs) / 1000)
+	if sleepMs == 0 {
+		return
+	}
+	logutil.BgLogger().Info(
+		"[pipelined dml] throttle",
+		zap.Int("sleepMs", sleepMs),
+		zap.Float64("writeThrottleRatio", txn.writeThrottleRatio),
+		zap.Uint64("session", txn.committer.sessionID),
+		zap.Uint64("startTS", txn.startTS),
+	)
```

Comment:
Logger util in client-go will not print arguments in reversed order.

```suggestion
	logutil.BgLogger().Info(
		"[pipelined dml] throttle",
		zap.Uint64("session", txn.committer.sessionID),
		zap.Uint64("startTS", txn.startTS),
		zap.Int("sleepMs", sleepMs),
		zap.Float64("writeThrottleRatio", txn.writeThrottleRatio),
	)
```